internal/integrationtest/util.go

```diff
@@ -0,0 +1,255 @@
+func (s TransportSpec) Test(t *testing.T) {
+	t.Run("reuse connection with round robin", s.TestReuseConnectionWithRoundRobin)
+	t.Run("backoff reconnection with round robin", s.TestBackoffWithRoundRobin)
+	t.Run("reconnect using round robin", s.TestReconnectWithRoundRobin)
+	t.Run("lose patience connecting with round robin", s.TestLosePatienceWithRoundRobin)
+}
+
+// NewClient returns a running dispatcher and a raw client for the echo
+// procedure.
+func (s TransportSpec) NewClient(t *testing.T, addrs []string) (*yarpc.Dispatcher, raw.Client) {
+	// Convert peer addresses into peer identifiers for a peer list.
+	ids := make([]peer.Identifier, len(addrs))
+	for i, addr := range addrs {
+		ids[i] = s.Identify(addr)
+	}
+
+	x := s.NewClientTransport(t)
+
+	pl := roundrobin.New(x)
+	pc := peerbind.Bind(pl, peerbind.BindPeers(ids))
+	ob := s.NewUnaryOutbound(x, pc)
+	d := yarpc.NewDispatcher(yarpc.Config{
+		Name: "client",
+		Outbounds: yarpc.Outbounds{
+			"service": transport.Outbounds{
+				ServiceName: "service",
+				Unary:       ob,
+			},
+		},
+	})
+	require.NoError(t, d.Start(), "start client dispatcher")
+	c := raw.New(d.ClientConfig("service"))
+	return d, c
+}
+
+// NewServer creates an echo server using the given inbound from any transport.
+func (s TransportSpec) NewServer(t *testing.T, addr string) (*yarpc.Dispatcher, string) {
+	x := s.NewServerTransport(t, addr)
+	ib := s.NewInbound(x, addr)
+
+	d := yarpc.NewDispatcher(yarpc.Config{
+		Name:     "service",
+		Inbounds: yarpc.Inbounds{ib},
+	})
+
+	handle := func(ctx context.Context, req []byte) ([]byte, error) {
+		return req, nil
+	}
+
+	d.Register(raw.Procedure("echo", handle))
+	require.NoError(t, d.Start(), "start server dispatcher")
+
+	return d, s.Addr(x, ib)
+}
+
+// TestLosePatienceWithRoundRobin is a test that any transport can apply to
+// exercise a transport dropping connections if the transport is stopped before
+// a pending request can complete.
+func (s TransportSpec) TestLosePatienceWithRoundRobin(t *testing.T) {
+	addr := "127.0.0.1:31172"
+	var wg sync.WaitGroup
+
+	client, c := s.NewClient(t, []string{addr})
+
+	wg.Add(1)
+	go func() {
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, 50*time.Millisecond)
+		defer cancel()
+		assert.Error(t, Call(ctx, c))
+		wg.Done()
+	}()
+
+	time.Sleep(10 * time.Millisecond)
+	client.Stop()
+
+	wg.Wait()
+}
+
+// TestReuseConnectionWithRoundRobin is a reusable test that any transport can
+// apply to cover connection reuse.
+func (s TransportSpec) TestReuseConnectionWithRoundRobin(t *testing.T) {
+	var wg sync.WaitGroup
+
+	server, addr := s.NewServer(t, ":0")
+	defer server.Stop()
+
+	client, c := s.NewClient(t, []string{addr})
+	defer client.Stop()
+
+	call := func() {
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, 50*time.Millisecond)
+		defer cancel()
+		assert.NoError(t, Call(ctx, c))
+		wg.Done()
+	}
+	wg.Add(10)
+	for i := 0; i < 10; i++ {
```

Comment:
I really don't like the arbitrary loops/sleeps/etc, at least #define them and document them